000-mm2-connector-offsets-support.md

```diff
@@ -0,0 +1,447 @@
+# MirrorMaker Connector Offsets Support
+
+Update the KafkaMirrorMaker2 resource to allow users to manage offsets
+
+Note: This proposal focuses on the KafkaMirrorMaker2 resource.
+Any reference to MirrorMaker is in regard to this resource, not the deprecated KafkaMirrorMaker resource.
+
+## Current situation
+
+[KIP-875][kip] added first class offsets support to Kafka Connect.
+This was done via four new endpoints:
+* GET /connectors/{connector}/offsets
+* PATCH /connectors/{connector}/offsets
+* DELETE /connectors/{connector}/offsets
+* PUT /connectors/{connector}/stop
+
+Support for the "stop" endpoint was added via [PR 9095](https://github.com/strimzi/strimzi-kafka-operator/pull/9095).
+Currently users that are utilising the KafkaMirrorMaker2 resource cannot make use of the API endpoints for managing offsets.
+
+Proposal [#076][proposal-76] discusses how to add support for these API endpoints to the KafkaConnector resource.
+However it did not cover the KafkaMirrorMaker2 resource.
```

Comment:
Offsets are actually only used by `MirrorSourceConnector`, so I wonder if we should only allow these operations on that connector?

Resetting/altering offsets of `MirrorCheckpointConnector` and `MirrorHeartbeatConnector` has no effect. If you try changing offsets on these connectors they still validate the request has a valid format.

Reply:
@mimaison if they aren't used why do the `MirrorCheckpointConnector` and `MirrorHeartbeatConnector` store offsets in Kafka?

Reply:
The Connect runtime automatically does that for source connectors. But these source offsets are not used by these 2 connectors to function.

Reply:
See https://github.com/apache/kafka/pull/14005 for more details.

Reply:
> @scholzj from an API perspective presumably the other connectors would inherit the listOffsets and alterOffsets properties? So I'm thinking we allow that but just remove the annotations without taking any action if the user requests list/alter/reset on a connector that isn't MirrorSourceConnector. Then if the mirrormaker-connector annotation includes both e.g. a MirrorSourceConnector and MirrorCheckpointConnector, we perform the action for MirrorSourceConnector but not MirrorCheckpointConnector. What do you think?

I would not be worried about the API I guess. The question is what to do if the user tries to work with those offsets:
1) Just reset them even if it doesn't influence anything?
2) Raise a warning and keep the annotation? (in which case you can also add a warning condition probably)
3) Raise the warning and remove the annotation? (in which case it is basically a warning in log only)

I do not have strong opinions for any of these options to be honest. I would probably go with 2) if I was doing it myself. But it is tight between the options for me, so I'm fine to take your path 😉

Reply:
After reading the comments on the PR that Mickael linked I'd be inclined for us to allow users to list/alter/reset offsets for all the MirrorMaker connectors. In the comments there is a discussion about whether users should be allowed to do these operations on MirrorCheckpoint and MirrorHeartbeat, or whether instead an UnsupportedOperationException should be thrown. The conclusion was that although the offsets aren't used by those connectors, there may still be reasons why a user would want to do those operations.

Given that Kafka allows it, I think we should also allow it. I'll update the proposal to capture this